### PR TITLE
Editable: Display a posts list suggestion in the link modal

### LINKED ISSUE #1040: Links: Search your old posts and pages in the input field
The old editor lets you search the link input field for old posts and pages you wrote. We should try and port this:

screen shot 2017-06-06 at 13 17 36

### PR DESCRIPTION
**Pull request by youknowriad**
closes #1040 

This is not ready but I'd appreciate some eyes to help me polish it.
The idea is to provide a dropdown of suggested posts when adding a link.

screen shot 2017-07-24 at 13 58 25

**Some notes:**

 - I'm only looking for posts, the issue suggests that we should look for pages too but unless I'm missing something the REST API can't look for both pages and posts. I might have to trigger two requests.
 - The search is triggered after typing at least 3 characters
 - This needs some Accessibility eyes, We can navigate using arrows but I guess this is far from perfect.

### CODE REVIEW

**Comment by aduth** [comment]
>unless I'm missing something the REST API can't look for both pages and posts

I think you're correct about this. Maybe worth flagging as "Needs API Endpoint" or otherwise some API conveniences? Two separate requests seems undesirable 😞

**Comment by jasmussen** [comment]
Nice, thank you for working on this! 

At a glance, this appears to be a nice UI that I'd thumb up. I can't really get it to work, though:

jul-24-2017 16-02-18

**Comment by youknowriad** [comment]
@jasmussen I suspect the same problems you encounter when performing API calls. Any JS error?

**Comment by jasmussen** [comment]
No JS errors, no. First time I tried, a little dialog did pop out though, and the spinner stopped. But nothing populated the dialog.

**Comment by youknowriad** [comment]
@nylen maybe you have some thoughts on the API issue here.

@jasmussen Unfortunately, I couldn't reproduce :(

**Comment by mtias** [comment]
@youknowriad works great for me on text nodes, but breaks the button on images:

image

**Comment by youknowriad** [comment]
@mtias Good catch, it should be better now (same for the button block)

**Comment by mtias** [comment]
@youknowriad cool, but it should open over the whole toolbar area :)

Master:
image

**Comment by youknowriad** [comment]
I force-pushed the last commit. I was missing a className change.

File: blocks/editable/format-toolbar/link-input.js [hunk 1]
```diff
@@ -0,0 +1,187 @@
+/**
```
Comment (mtias) [inline blocks/editable/format-toolbar/link-input.js]: It is a bit weird that link-input is under `format-toolbar`, considering the link input can be used inline or block level, what about placing it together with `url-input`? The current url-input/index.js could be renamed `button.js` maybe.

File: blocks/editable/format-toolbar/link-input.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+			this.suggestionNodes[ index ] = ref;
+		};
+	}
+
+	updateSuggestions( value ) {
+		if ( this.suggestionsRequest ) {
+			this.suggestionsRequest.abort();
+		}
+
+		if ( value.length < 2 ) {
```
Comment (mtias) [inline blocks/editable/format-toolbar/link-input.js]: A comment could help here.

**Comment by mtias** [comment]
Just a comment on the component structure, but this looks good to me.

File: blocks/editable/format-toolbar/link-input.js (hunk 1 above)
Comment (youknowriad) [inline blocks/editable/format-toolbar/link-input.js]: 👍 I like this proposal, it's there because it was extracted from the `format-toolbar` at first.

**Review by mtias — APPROVED** [review]
This looks good to me. 🚢

**Comment by nylen** [comment]
The API issue here is tracked at https://core.trac.wordpress.org/ticket/39965.

**Comment by afercia** [comment]
> I'm only looking for posts, the issue suggests that we should look for pages too but unless I'm missing something the REST API can't look for both pages and posts. I might have to trigger two requests.

The current functionality in WP searches also pages and also CPTs. See
`wp_ajax_wp_link_ajax()`
`_WP_Editors::wp_link_query()`

Worth also considering the REST API returns by default _all_ the post fields including the content, and the response could be huge in size. Even several hundreds KB or more than 1MB with very long posts. See related considerations on:
https://core.trac.wordpress.org/ticket/38920
https://core.trac.wordpress.org/ticket/38922
and especially:
"REST API: Provide interface to include or exclude specific fields from response JSON"
https://core.trac.wordpress.org/ticket/38131

I think these concerns should definitely go in a new issue.

**Pull request merged**